authorizations/message-filter/src/contract.rs

```diff
@@ -0,0 +1,374 @@
+
+const CONTRACT_NAME: &str = "crates.io:whitelist";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+#[serde(untagged)]
+pub enum Value {
+    Null,
+    Bool(bool),
+    U128(Uint128),
+    Number(u64), // Why doesn't u128 work here?
+    String(String),
+    Array(Vec<Value>),
+    Object(Map<String, Value>),
+}
+
+fn msg_to_value(msg: &CosmosMsg) -> Result<Value, ContractError> {
+    let serialized = to_string(msg).map_err(|_| ContractError::CustomError {
+        val: "invalid CosmosMsg".to_string(),
+    })?;
+
+    str_to_value(&serialized)
+}
+
+fn str_to_value(msg: &str) -> Result<Value, ContractError> {
+    from_str(msg).map_err(|_| ContractError::CustomError {
+        val: "invalid str".to_string(),
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let config = Config {
+        dao: msg.dao,
+        kind: msg.kind,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default().add_attribute("action", "instantiate"))
+}
+
+fn deep_partial_match(msg: &Value, authorization: &Value) -> bool {
+    match authorization {
+        Value::Object(auth_map) => {
+            if auth_map.is_empty() {
+                return true;
+            }
+
+            let mut matching = true;
+            if let Value::Object(msg_map) = msg {
+                for (key, val) in auth_map {
+                    if !msg_map.contains_key(key) {
+                        return false;
+                    };
+                    matching = matching & deep_partial_match(msg_map.get(key).unwrap(), val);
+                }
+            } else {
+                return false;
+            }
+            matching
+        }
+        Value::Array(auth_array) => {
+            // Comparing arrays manually because PartialEq doesn't understand use our deep matching.
+            let mut matching = true;
+            if let Value::Array(msg_array) = msg {
+                if msg_array.len() != auth_array.len() {
+                    return false;
+                }
+                for (i, elem) in auth_array.iter().enumerate() {
+                    matching = matching & deep_partial_match(&msg_array[i], &elem);
+                }
+            } else {
+                return false;
+            }
+            matching
+        }
+        _ => authorization == msg,
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::AddAuthorization { addr, msg } => {
+            let config = CONFIG.load(deps.storage)?;
```

Comment:
can we make these blocks into their own functions?

Reply:
done!